  <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head profile="http://gmpg.org/xfn/11">
<title>Bruno P. Kinoshita &mdash; Bioinformatics tools: Stacks</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="description" content="Bruno de Paula Kinoshita personal web site" />
<meta name="keywords"
    content="kinoshita, bruno de paula kinoshita, geek, geek stuff, java, python, c, c++, msn now playing, neural network, mackenzie, sao paulo, brasil, brazil, nerd, nerd stuff, cool, jenkins, illustration, quadrinhos, machine learning, linguistic, apache" />
<meta name="author" content="Bruno P. Kinoshita" />
<meta name="generator" content="PieCrust 1.2.0" />
<meta name="template-engine" content="Twig" />
<meta name="robots" content="all=index,follow" />
<meta name="revisit-after" content="15 days" />
<meta http-equiv="Pragma" content="no-cache" />
<meta name="distribution" content="Global" />
<meta name="url" content="http://kinoshita.eti.br/" />
<meta name="copyright" content="Bruno de Paula Kinoshita" />
<meta name="Googlebot" content="all" />
<meta name="rating" content="general" />
<meta http-equiv="expires" content="0" />
<meta name="city" content="Sao Paulo" />
<meta name="state" content="Brasil, Sao Paulo" />
<meta http-equiv="content-language" content="en" />
<meta http-equiv="imagetoolbar" content="no" />
<meta id="page_favicon" href="http://kinoshita.eti.br/favicon.ico" rel="icon" type="image/x-icon" />
<link rel="shortcut icon" type="image/x-icon" href="http://kinoshita.eti.br/favicon.ico">
<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
<meta name="viewport" content="width=device-width,initial-scale=1">
<!--[if lt IE 9]>
<script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
<script src="http://kinoshita.eti.br/js/jquery-1.6.1.min.js"  type="text/javascript" charset="utf-8"></script>
<script src="http://kinoshita.eti.br/js/jquery.prettyPhoto.js"></script>
<script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<link rel="alternate" type="application/rss+xml" title="Bruno P. Kinoshita &raquo; Feed" href="http://kinoshita.eti.br//feed.xml" />
<link rel="stylesheet" href="http://kinoshita.eti.br/css/prettyPhoto.css" type="text/css" media="screen, projection" />
<link rel="stylesheet" href="http://kinoshita.eti.br/css/bootstrap.css" type="text/css" media="screen, projection" />
<link rel="stylesheet" href="http://kinoshita.eti.br/css/bootstrap-responsive.css" type="text/css" media="screen, projection" />
<link rel="stylesheet" href="http://kinoshita.eti.br/css/style.css" type="text/css" media="screen, projection" />
<!--[if lt IE 8]>
    <link rel="stylesheet" href="/css/blueprint/ie.css" type="text/css" media="screen, projection" />
<![endif]-->
<link rel="stylesheet" type="text/css" href="http://fonts.googleapis.com/css?family=Droid+Sans">
 <link rel="stylesheet" href="/css/default.css" type="text/css" media="all" />
 
</head>
<body class="home blog">
    <div class="container-fluid">
            <div class="row-fluid">
            <div id="left" class="span2">
                <div style="padding-top: 6px;">
                    <ul>
                        <li><a style="color: #E13F4D;" class="mainlink"
                            href="http://kinoshita.eti.br/">Kinoshita</a></li>
                    </ul>
                </div>
                <div class="menu-main-menu-container">
    <ul
        id="menu-main-menu"
        class="menu">
        <li
            class="menu-item"><a
            href="http://kinoshita.eti.br/software.html">Software</a></li>
        <li
            class="menu-item"><a
            href="http://kinoshita.eti.br/illustrations.html">Illustrations</a></li>
        <li
            class="menu-item"><a
            href="http://kinoshita.eti.br/articles.html">Articles</a></li>
        <li
            class="menu-item"><a
            href="http://kinoshita.eti.br/about.html">About &#038; CV</a></li>
        <li
            class="menu-item"><a
            href="http://kinoshita.eti.br/contact.html">Contact</a></li>
    </ul>
</div>                <hr class="space" />
                <img src="http://kinoshita.eti.br/img/bruno.jpg"
                    title="Hmmmm" alt="Hmmmm" width="134px" height="215px" />
                <hr class="space" />
                <div class="logos">
    <a href="http://www.tupilabs.com"><img
        src="http://kinoshita.eti.br/img/logos/tupilabs_logo.png"
        title="TupiLabs"
        alt="TupiLabs" /></a> <a href="http://www.javacodegeeks.com/"><img
        src="http://kinoshita.eti.br/img/logos/logo_jcg_115x100.png"
        title="Java Code Geeks"
        alt="Java Code Geeks" /></a> <a href="http://www.open-bio.org/wiki/BOSC_2012"><img
        src="http://kinoshita.eti.br/img/logos/ISMB2012.jpg"
        title="BOSC 2012"
        alt="BOSC 2012" /></a> <a href="http://www.stpcon.com/"><img
        src="http://kinoshita.eti.br/img/logos/stpconspring2012.png"
        title="STPCon Spring 2012"
        alt="STPCon Spring 2012" /></a> <a href="http://www.oracle.com/javaone/lad-pt/"><img
        src="http://kinoshita.eti.br/img/logos/javaone.png"
        title="JavaOne 2011 LAD"
        alt="JavaOne 2011 LAD" /></a> <a href="http://www.jcertif.com"><img
        src="http://kinoshita.eti.br/img/logos/JCertif_Conf2011_2.png"
        title="JCertif 2011"
        alt="JCertif 2011" /></a> <a href="http://www.thedevelopersconference.com.br/tdc/2011/"><img
        src="http://kinoshita.eti.br/img/logos/banner-TDC2011-234x60.png"
        title="The Developer's Conference 2011"
        alt="The Developer's Conference 2011" /></a> <a href="http://belgiumtestingdays.com/past.php"><img
        src="http://kinoshita.eti.br/img/logos/btd2011_speakers_banner_1.png"
        title="Belgium Testing Days 2011"
        alt="Belgium Testing Days 2011" /></a> <a href="http://www.encontroagil.com.br/2010/principal/home.html"><img
        src="http://kinoshita.eti.br/img/logos/banner_encontro_agil.png"
        title="Encontro &Aacute;gil 2010"
        alt="Encontro &Aacute;gil 2010" /></a>
</div>            </div>
                        <div id="right" class="span8">
                <div id="main">
                <div class="post type-post status-publish format-standard hentry">
    <h1 class='post-topic'>Bioinformatics tools: Stacks</h1>
    <small>Bruno P. Kinoshita</small> @
	<small>September 25, 2012&nbsp;16:08:03</small>
    <div class="entry">
       <a href="https://twitter.com/share" class="twitter-share-button" data-url="http://kinoshita.eti.br/2012/09/25/bioinformatic-tools-series-stacks.html" data-text="Bioinformatics tools: Stacks" data-via="kinow" data-hashtags="opensource">Tweet</a>
<script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');</script>

	   <p><em>It is the first post about <strong>bioinformatics tools</strong>, but I will try to post more about other tools such as MrBayes, Structure, maybe some next generation sequencing tools too, and Bioperl, Biojava, and so on.</em></p>

<p><em>As I am more a computer geek, rather than a bioinformatics one, I will focus on requirements for running these tools on clusters and the requirements to install them on your machine. The instructions require that you have an intermediary knowledge on *nix OS and sometimes a bit of programming experience.</em></p>

<p><em>I will be using tutorials available on the Internet and hosting my code in <a href="http://www.github.com/kinow" alt="GitHub kinow" title="GitHub kinow">GitHub/kinow</a>. Hammer time!</em></p>

<!-- more -->

<hr />

<p style="text-align: center"><a href="/2012/09/25/bioinformatic-tools-series-stacks/stacks_logo_300_156.png"><img src="/2012/09/25/bioinformatic-tools-series-stacks/stacks_logo_300_156.png" alt="" title="stacks_logo" width="300" height="156" class="aligncenter size-medium wp-image-1058" /></a></p>

<p><strong>Stacks</strong> is hosted at <a href="http://www.uoregon.edu/" alt="Oregon University" title="Oregon University">Oregon University</a>, so if you are googling about it right now, probably a query including both &#8216;stacks&#8217; and &#8216;oregon&#8217; may give you better results (it&#8217;s good to know when you forget the web address). This is the stacks homepage: <a href="http://creskolab.uoregon.edu/stacks/">http://creskolab.uoregon.edu/stacks/</a>.</p>

<blockquote cite="http://creskolab.uoregon.edu/stacks/">&#8220;Stacks is a software pipeline for building loci out of a set of short-read sequenced samples. Stacks was developed for the purpose of building genetic maps from RAD-Tag Illumina sequence data, but can also be readily applied to population studies, and phylogeography.&#8221;</blockquote>

<p>I will follow the tutorial <a href="http://creskolab.uoregon.edu/stacks/pe_tut.php" alt="Stacks tutorial" title="Stacks tutorial">&#8220;building mini-contigs from paired-end sequences&#8221;</a> to demonstrate how to use Stacks, and for each step executed I&#8217;ll post the directory and database sizes, as well as the processing time. My hardware is an Intel i5 quad core 2.30 GHz with 6GB of memory, running Debian 6.</p>

<!--more-->

<h2>Installing on your local computer</h2>

<p>The installation is quite simple. However, Stacks come with a Web application, and during its pipeline analysis sometimes it uploads data to a database. So in our set up I&#8217;m going to use an Apache HTTP web server and MySQL database. Here&#8217;s a list of things you should have installed in your machine: </p>

<ul>
<li><a href="http://httpd.apache.org" title="Apache httpd">Apache HTTP web server</a> (you could try NGINX, or others too, but you would need a PHP interpreter) - my version: Apache/2.2.22 (Debian)</li>
<li><a href="http://www.php.net" title="PHP">PHP</a> (not sure if 5> is required) - my version: Ver 14.14 Distrib 5.5.24, for debian-linux-gnu (x86_64) using readline 6.2</li>
<li>PHP MDB2 module installed (pear install MDB2 does the trick most of the times)</li>
<li><a href="http://www.mysql.org/" title="MySQL">MySQL</a> database (stacks come with a SQL script for MySQL, not sure how hard it would be to adapt this script and the perl utilities to use Oracle, DB2, etc) - my version: Ver 14.14 Distrib 5.5.24, for debian-linux-gnu (x86_64) using readline 6.2</li>
<li><a href="http://www.perl.org" title="Perl">Perl</a> (5 is fine, didn&#8217;t test with 6) - This is perl 5, version 14, subversion 2 (v5.14.2) built for x86_64-linux-gnu-thread-multi</li>
<li>DBD::mysql and SPREADSHEET::WriteExcel modules (<a href="http://www.cpan.org" title="CPAN">CPAN</a>, yay!)</li>
<li><a href="http://www.ebi.ac.uk/~zerbino/velvet/" title="Velvet">Velvet</a> (for contigs) - my version: 1.2.07</li>
<li>A gcc compiler with OpenMP for multi-threading. My version: 
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.7/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v &#8212;with-pkgversion=&#8217;Debian 4.7.1-8&#8217; &#8212;with-bugurl=file:///usr/share/doc/gcc-4.7/README.Bugs &#8212;enable-languages=c,c++,go,fortran,objc,obj-c++ &#8212;prefix=/usr &#8212;program-suffix=-4.7 &#8212;enable-shared &#8212;enable-linker-build-id &#8212;with-system-zlib &#8212;libexecdir=/usr/lib &#8212;without-included-gettext &#8212;enable-threads=posix &#8212;with-gxx-include-dir=/usr/include/c++/4.7 &#8212;libdir=/usr/lib &#8212;enable-nls &#8212;with-sysroot=/ &#8212;enable-clocale=gnu &#8212;enable-libstdcxx-debug &#8212;enable-libstdcxx-time=yes &#8212;enable-gnu-unique-object &#8212;enable-plugin &#8212;enable-objc-gc &#8212;with-arch-32=i586 &#8212;with-tune=generic &#8212;enable-checking=release &#8212;build=x86_64-linux-gnu &#8212;host=x86_64-linux-gnu &#8212;target=x86_64-linux-gnu
Thread model: posix
gcc version 4.7.1 (Debian 4.7.1-8)</li>
</ul>

<p>Install Stacks with the classic <em>./configure</em>, <em>make</em>, <em>make install</em>. It will put several utilities and Perl scripts in your PATH. For testing if the installation worked, try running <em>ustacks</em>. You may need to update your Apache HTTP settings to include a virtual directory to the stacks web folder. In my computer this folder is located at <em>/usr/local/share/stacks/php</em>.</p>

<h2>Installing on your cluster</h2>

<p>Stacks uses <strong>OpenMP</strong> for running over multiple CPU&#8217;s. I have a small commodity-hardware cluster that is quite handy for MapReduce and MPI processing, but for this case I will have to use only one machine. So I&#8217;m using my notebook with four cores, and using <a href="http://ganglia.sourceforge.net" title="Ganglia">Ganglia</a> and htop/free for monitoring CPU and memory usage.</p>

<h2>Following the Stacks tutorial</h2>

<p>The tutorial requires that you download samples and create some directories and a database. Here&#8217;s the status of my system before the analysis.</p>

<p><strong>Database size: 880KB</strong></p>

<p><p><strong>Directory size: 2.6 GB</strong></p>
<code>kinow@chuva:~/Desktop/pe_tut$ du -h
4.0K    ./stacks
4.0K    ./assembled
2.6G    ./samples
4.0K    ./raw
4.0K    ./pairs
2.6G    .</code></p>

<p>After these steps, you start your pipeline with the <em>denovo_map.pl</em> script. The denovo_map.pl script took 23 minutes to finish, and here&#8217;s the new status of my system.</p>

<p><strong>Database size: 1.7 GB</strong></p>

<p><p><strong>Directory size: 3.5 GB</strong></p>
<code>kinow@chuva:~/Desktop/pe_tut$ du -h
971M    ./stacks
4.0K    ./assembled
2.6G    ./samples
4.0K    ./raw
4.0K    ./pairs
3.5G    .</code></p>

<p>The next step is the <em>export_sql.pl</em> script. It took 31 minutes to finish.</p>

<p><strong>Database size: 1.7 GB</strong></p>

<p><p><strong>Directory size: 3.5 GB</strong></p>
<code>kinow@chuva:~/Desktop/pe_tut$ du -h
971M    ./stacks
4.0K    ./assembled
2.6G    ./samples
4.0K    ./raw
4.0K    ./pairs
3.5G    .</code></p>

<p>Half way to go. Now execute the <em>sort_read_pairs.pl</em> script. This was the longest and the one that consumed more memory, it took 49 minutes to finish.</p>

<p><strong>Database size: 1.7 GB</strong></p>

<p><p><strong>Directory size: 3.7 GB</strong></p>
<code>kinow@chuva:~/Desktop/pe_tut$ du -h
971M    ./stacks
4.0K    ./assembled
2.6G    ./samples
4.0K    ./raw
170M    ./pairs
3.7G    .</code></p>

<p>It&#8217;s time to get the contigs with Velvet now with <em>exec_velvet.pl</em>. Don&#8217;t worry, this one is faster, only 15 minutes.</p>

<p><strong>Database size: 1.7 GB</strong></p>

<p><p><strong>Directory size: 3.7 GB</strong></p>
<code>kinow@chuva:~/Desktop/pe_tut$ du -h
971M    ./stacks
3.4M    ./assembled
2.6G    ./samples
4.0K    ./raw
170M    ./pairs
3.7G    .</code></p>

<p>Last two scripts to execute. First <em>load_sequences.pl</em>, that apparently loads some data into the database. It took 3 minutes to finish.</p>

<p><strong>Database size: 1.8 GB</strong></p>

<p><p><strong>Directory size: 3.7 GB</strong></p>
<code>kinow@chuva:~/Desktop/pe_tut$ du -h
971M    ./stacks
3.4M    ./assembled
2.6G    ./samples
4.0K    ./raw
170M    ./pairs
3.7G    .</code></p>

<p>And finally <em>index_radtags.pl</em> which took only 1 minute.</p>

<p><strong>Database size: 1.8 GB</strong></p>

<p><p><strong>Directory size: 3.7 GB</strong></p>
<code>kinow@chuva:~/Desktop/pe_tut$ du -h
971M    ./stacks
3.4M    ./assembled
2.6G    ./samples
4.0K    ./raw
170M    ./pairs
3.7G    .</code></p>

<h2>CPU, memory and disk usage</h2>

<p>I was using Ganglia for monitoring Hadoop jobs, so I saved the graphs for the timeframe of the pipeline execution. I started to execute this analysis at 10AM, and finished about 2 and a half hours later (the process is not automated, so sometimes I had to check the output and documentation before executing the next step).</p>

<p>Following you&#8217;ll find the raw graphs and another version with legends to help understand what was being executed.</p>

<h3>CPU usage</h3>

<p style="text-align: center"><a href="/2012/09/25/bioinformatic-tools-series-stacks/cpu.png"><img src="/2012/09/25/bioinformatic-tools-series-stacks/cpu.png" alt="" title="CPU usage" width="747" height="418" class="aligncenter size-full wp-image-1043" /></a></p>

<p style="text-align: center"><a href="/2012/09/25/bioinformatic-tools-series-stacks/cpu_legends.png"><img src="/2012/09/25/bioinformatic-tools-series-stacks/cpu_legends.png" alt="" title="CPU usage with legends" width="747" height="418" class="aligncenter size-full wp-image-1050" /></a></p>

<h3>Memory usage</h3>

<p style="text-align: center"><a href="/2012/09/25/bioinformatic-tools-series-stacks/memory.png"><img src="/2012/09/25/bioinformatic-tools-series-stacks/memory.png" alt="" title="Memory usage" width="747" height="430" class="aligncenter size-full wp-image-1048" /></a></p>

<p style="text-align: center"><a href="/2012/09/25/bioinformatic-tools-series-stacks/memory_legends.png"><img src="/2012/09/25/bioinformatic-tools-series-stacks/memory_legends.png" alt="" title="Memory usage with legends" width="747" height="430" class="aligncenter size-full wp-image-1052" /></a></p>

<h3>Disk usage</h3>

<p>I thought the disk graphs from Ganglia wouldn&#8217;t represent the disk usage very well, specially since my disk has about 500GB, so it&#8217;s harder to see the changes. So I used some R to plot a graph that I hope can demonstrate the usage for this tutorial. <strong>It is important to highlight that we are using samples from a tutorial, and your analysis may produce high or lower disk usage.</strong></p>

<p style="text-align: center"><a href="/2012/09/25/bioinformatic-tools-series-stacks/disk.png"><img src="/2012/09/25/bioinformatic-tools-series-stacks/disk.png" alt="" title="Disk Usage" width="824" height="630" class="aligncenter size-full wp-image-1055" /></a></p>

<p>Hope it helps you using Stacks. Kudos to all developers and contributors of this great tool.</p>

<pre class="r" style="font-family:monospace;">disk=c(2.6,3.5,3.5,3.7,3.7,3.7,3.7)
db=c(0.0008,1.7,1.7,1.7,1.7,1.8,1.8)
x=(1:7)
plot(x=x, xlab=&quot;Tutorial Step (1 to 7)&quot;, y=disk, ylab=&quot;Disk Usage (GB)&quot;, col=&quot;red&quot;, type=&quot;o&quot;,ylim=c(0,6),xlim=c(0,8))
lines(x=x, y=db, type=&quot;o&quot;, pch=22, lty=2, col=&quot;blue&quot;)
title(main=&quot;Disk Usage&quot;, font.main=2)
abline(v=c(0:8), col=&quot;grey40&quot;, lty=&quot;dotted&quot;)
abline(h=c(0:5), col=&quot;grey10&quot;, lty=&quot;dotted&quot;)
df = data.frame(disk,db)
totals=rowSums(df,na.rm=T)
lines(x=x, y=totals, type=&quot;o&quot;, pch=22, lty=2, col=&quot;green&quot;) 
legend(0,6,c(&quot;DB&quot;, &quot;Disk&quot;, &quot;Total&quot;), cex=0.6,col=c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;),pch=21:23,lty=1,y.intersp=0.2,x.intersp=0.2,pt.lwd=1,adj=0,inset=c(0.5,0.5))</pre>

<h4>References</h4>

<ul>
<li>Stacks - <a href="http://creskolab.uoregon.edu/stacks/">http://creskolab.uoregon.edu/stacks/</a></li>
<li>Tutorial: building mini-contigs from paired-end sequences - <a href="http://creskolab.uoregon.edu/stacks/pe_tut.php">http://creskolab.uoregon.edu/stacks/pe_tut.php</a></li>
</ul>

    </div>
    <p class="postmetadata">
                <small>tags          [&nbsp;<a
            href="/tag/stacks.html"
            rel="tag">stacks</a>&nbsp;]&nbsp;
                 [&nbsp;<a
            href="/tag/bioinformatics.html"
            rel="tag">bioinformatics</a>&nbsp;]&nbsp;
                 [&nbsp;<a
            href="/tag/ngs.html"
            rel="tag">ngs</a>&nbsp;]&nbsp;
                 [&nbsp;<a
            href="/tag/dna-sequencing.html"
            rel="tag">dna sequencing</a>&nbsp;]&nbsp;
                </small>
                        <br/>
        <small>posted in 
        [&nbsp;<a
            href="/bioinformatics.html"
            title="View all posts in bioinformatics"
            rel="category tag">bioinformatics</a>&nbsp;]&nbsp;
        category </small>
            </p>
    <div id="navigation">
                    &larr;&nbsp;<a href="/2012/09/25/paper-patterns-for-introducing-a-superclass-for-test-classes.html" id="navigation-next">Previous post</a>
                            &mdash; <a href="/2012/09/20/running-word-count-example-on-a-hadoop-commodity-hardware-cluster-and-on-a-hadoop-local-installation.html" id="navigation-next">Next Entries</a>&nbsp;&rarr;
            </div>
</div>
                    <br/>
                </div><!-- /#main -->
                            <div id="copy" class="">
                <p>&copy; <a href="http://kinoshita.eti.br/about"><span
                            class="b">Bru</span><span
                            class="r">no de</span> <span
                            class="a">Paula</span> <span
                            class="s"></span><span
                            class="i">Kinos</span><span
                            class="l">hita</span></a>
                </p>
            </div>
 
                                
            </div><!-- /#right -->
        </div>
    </div>
</body>
</html>
<script type="text/javascript" charset="utf-8">
  $(document).ready(function(){
    $("a[rel^='prettyPhoto']").prettyPhoto({
    	social_tools: false,
    	autoplay_slideshow: false, 
    	hideflash: true, 
    	autoplay: false, 
    	overlay_gallery: false
    });
  });
</script>
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-32767310-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
